### PR TITLE
Desul atomics with SYCL

### PR DESCRIPTION
**Pull request by masterleinad**
Based on top of https://github.com/kokkos/kokkos/pull/3247. As discussed with @crtrott we want to merge #3247 first.

### CODE REVIEW

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Retest this please.

**Comment by masterleinad** [comment]
Depends on https://github.com/desul/desul/pull/21.

**Comment by masterleinad** [comment]
We decided to first merge the desul changes here and only after that upstream.

**Comment by joeatodd** [comment]
I've tested this with the SYCL backend (Intel's) using LAMMPS. This is a significant performance boost over the previous non-native atomic code (like 3x faster!). Big thumbs up from me :)

File: core/src/desul/atomics/Compare_Exchange_SYCL.hpp [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+namespace desul {
+
+template<class MemoryOrder, class MemoryScope>
+inline void atomic_thread_fence(MemoryOrder, MemoryScope) {
+  DESUL_SYCL_NAMESPACE::atomic_fence(DesulToSYCLMemoryOrder<MemoryOrder>::value,
+                                     DesulToSYCLMemoryScope<MemoryScope>::value);
+}
+
+template <typename T, class MemoryOrder, class MemoryScope>
+typename std::enable_if<sizeof(T) == 4, T>::type atomic_compare_exchange(
```
Comment (nliber) [inline core/src/desul/atomics/Compare_Exchange_SYCL.hpp:27]: Do we also want to SFINAE or static_assert that T is trivially copyable?

**Review by nliber — APPROVED** [review]

File: core/src/desul/atomics/Compare_Exchange_SYCL.hpp (hunk 1 above)
Comment (crtrott) [reply core/src/desul/atomics/Compare_Exchange_SYCL.hpp:27]: we don't really do it on any of the other ones, lets do that separately in the post "Now Kokkos Is Working Cleanup" in the desul repo.

**Review by crtrott — APPROVED** [review]

**Pull request merged**